### PR TITLE
Rewrite to xPath and bug fixes - Version 3.0.0 (?beta)

### PR DESCRIPTION
**Pull request by BeroBurny**
Rewrite inspired by issue #88

### TODO  
- [x] #88 
- [ ] #75
- [ ] #79 
  - [x] options
  - [ ] ?? handle edge case
  - [ ] write test for all cases
- [ ] #86 
- [ ] test it if is resolved with xPath #89 
- [ ] #93 
- [ ] #94 

There will be breaking changes, and this is a perfect time to suggest your own ideas to implement.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -56,7 +67,7 @@ export const RECOMMENDED_METAMASK_VERSION = 'v10.8.1';
 export async function launch(puppeteerLib: typeof puppeteer, options: LaunchOptions): Promise<puppeteer.Browser> {
   if (!options || !options.metamaskVersion)
     throw new Error(
-      `Pleas provide "metamaskVersion" (use recommended "${RECOMMENDED_METAMASK_VERSION}" or "latest" to always get latest release of MetaMask)`,
+      `Pleas provide "metamaskVersion" (recommended "${RECOMMENDED_METAMASK_VERSION}" or "latest" to always get latest release of MetaMask)`,
```
[the remark this answers is not in the document]
Comment (BeroBurny) [reply src/index.ts:73]: this error is shown when is not present a metamask version from options, it is not related to what we talked about a few days ago

File: src/helpers/selectors.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import { ElementHandle, Page } from 'puppeteer';
+
+export const getElementByContent = (page: Page, text: string, type = '*'): Promise<ElementHandle | null> =>
+  page.waitForXPath(`//${type}[contains(text(), '${text}')]`);
```
Comment (scottcanoni) [inline src/helpers/selectors.ts:4]: XPath is great for looking for elements with text inside.  You might consider switching `text()` for `.`

`page.waitForXPath(`//${type}[contains(., '${text}')]`);`

In my opinion, `text()` reads more clearly but in case the HTML structure changes, you might want the `.`.  I am curious, what do you think?

https://stackoverflow.com/a/41862373/1026017

**Comment by keithfrazier98** [comment]
I do have one idea, which I am not sure how other people are currently overcoming the issue it stems 
from. Currently,  I am not sure how to prompt metamask to show the connection screen below past the first time I access metamask with my dapp.

Screen Shot 2022-03-03 at 12 31 21 PM

In a normal circumstance, there is either a popup in the metamask widget, or if the user has that setting disabled, there is a button that says 'not connected' as seen below.
Screen Shot 2022-03-03 at 12 20 02 PM

And you can click that button to then access the connection popup in the widget and connect to a site, but this option isn't available in the page view of metamask, and the popup doesn't appear in the page view either. Considering that the metamask popup widget is closed automatically, I had to find a new way to connect more than one account in a wallet to a site. 

I settled upon importing both wallets at the beginning of the settion (using importPk) then writing this function to connect all accounts to the dapp when the connection window appears for the first time. 

```
export async function approve(page: puppeteer.Page, selectAll: boolean = false, version?: string): Promise<void> {
  await page.bringToFront();
  await page.reload();

  try {
    if (selectAll) {
      const checkbox = await page.waitForSelector(".permissions-connect-choose-account__select-all > input", {
        timeout: 3000,
      });
      if (checkbox) await checkbox.click({ clickCount: 2 });
    }
  } catch (error) {
    console.log("Couldnt select all, maybe there is only one available.");
  }

  const button = await page.waitForSelector("button.button.btn-primary", { timeout: 3000 });
  if (button) await button.click();

  const connectButton = await page.waitForSelector("button.button.btn-primary", { timeout: 3000 });
  if (connectButton) await connectButton.click();
}
```

Which really it's just a modification to the approve function from dappeteer, but instead of connecting the first account, there is on optional param that allows you to click the select all button to connect all accounts at once. 
This would be a useful feature out the box for dappeteer.

**Comment by wtesler** [comment]
Does this fix https://github.com/ChainSafe/dappeteer/issues/102 ? I saw another issue with the same error, I believe this draft was marked as a fix for it. What is the underlying issue there?

**Comment by wtesler** [comment]
Does this fix https://github.com/ChainSafe/dappeteer/issues/102 ? I saw another issue with the same error, I believe this draft was marked as a fix for it. What is the underlying issue there?

File: README.md [hunk 3]
```diff
@@ -16,8 +16,7 @@ import puppeteer from 'puppeteer';
 import dappeteer from '@chainsafe/dappeteer';
 
 async function main() {
-  const browser = await dappeteer.launch(puppeteer, { metamaskVersion: 'v10.8.1' });
-  const metamask = await dappeteer.setupMetamask(browser);
+  const [metamask, page] = await dappeteer.bootstrap(puppeteer, { metamaskVersion: 'v10.10.1' });
```
Comment (mpetrunic) [inline README.md:19]: 10.15.0?

File: package.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@chainsafe/dappeteer",
-  "version": "2.4.1",
+  "version": "3.0.0-next.0",
```
Comment (mpetrunic) [inline package.json:3]:
```suggestion
  "version": "3.0.0-rc.0",
```

File: package.json [hunk 5]
```diff
@@ -14,7 +14,8 @@
     "build": "tsc -p tsconfig.build.json",
     "lint": "eslint --color --ext .ts src/ test/",
     "lint:fix": "yarn run lint --fix",
-    "test": "mocha"
+    "test": "mocha",
+    "test:dev": "mocha --timeout 36000000"
```
Comment (mpetrunic) [inline package.json:18]: do we even need this command?

File: package.json [hunk 6]
```diff
@@ -52,14 +54,14 @@
     "jest-environment-node": "^27.1.1",
     "mocha": "7.2.0",
     "prettier": "^2.2.1",
-    "puppeteer": "5.5.0",
+    "puppeteer": "14.0.0",
     "serve-handler": "5.0.8",
     "solc": "0.5.2",
     "ts-node": "9.1.1",
-    "typescript": "^3.8",
+    "typescript": "^4.1",
```
Comment (mpetrunic) [inline package.json:61]: maybe 4.7?

File: package.json [hunk 7]
```diff
@@ -52,14 +54,14 @@
...
     "web3": "1.3.4"
   },
   "peerDependencies": {
-    "puppeteer": ">5"
+    "puppeteer": ">14"
```
Comment (mpetrunic) [inline package.json:65]: can we relax this condition a bit?

File: src/types.ts [hunk 8]
```diff
@@ -0,0 +1,59 @@
+import { BrowserLaunchArgumentOptions, Page } from 'puppeteer';
+
+import { Path } from './setup/metamaskDownloader';
+
+export type LaunchOptions = OfficialOptions | CustomOptions;
+
+export type OfficialOptions = BrowserLaunchArgumentOptions & {
+  metamaskVersion: 'v10.10.1' | 'latest' | string;
```
Comment (mpetrunic) [inline src/types.ts:8]:
```suggestion
  metamaskVersion: typeof RECOMMENDED_METAMASK_VERSION | 'latest' | string;
```

File: src/types.ts [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+  metamaskLocation?: Path;
+};
+
+export type CustomOptions = BrowserLaunchArgumentOptions & {
+  metamaskVersion?: undefined;
```
Comment (mpetrunic) [inline src/types.ts:13]:
```suggestion
  metamaskVersion?: string;
```

File: package.json (hunk 5 above)
Comment (BeroBurny) [reply package.json:18]: There is no need but it is pretty helpful for debugging metamask UI flow

File: package.json (hunk 4 above)
Comment (BeroBurny) [reply package.json:3]: I don't think it is ready for release (and for not having breaking changes) but it is in better shape than 2xx that is the rationale behind `beta`, should we still process with `rc`?

Comment (mpetrunic) [reply package.json:3]: yeah, it's release candidate, doesn't mean it's ready to be promoted
